### DISCUSSION TITLE
best practice for "inactive" subdomains?

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#28092] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Dear Moose Community,

For various tasks in the field of engineering and mining, the excavation (and possibly refilling) of defined volumes must be modelled. In Moose this is usually to be done by "parking" elements in "inactive subdomains" via one of the SubdomainModifiers (e.g  [CoupledVarThresholdElementSubdomainModifier](https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html) or [TimedSubdomainModifier](https://mooseframework.inl.gov/source/userobjects/TimedSubdomainModifier.html)). Initially inactive subdomains may be created via add_subdomain_ids and/or add_subdomain_names in [MooseMesh](https://mooseframework.inl.gov/source/mesh/MooseMesh.html).

With this discussion entry I'd like to gain insight on the best practice on how to define the inactive subdomain.

As recommented [here](https://github.com/idaholab/moose/discussions/26079#discussioncomment-7623782) and shown [there](https://github.com/idaholab/moose/discussions/26941) for the "inactive" subdomains a Diffusion or MatDiffusion kernels may be used. As stated in the first link, `kernel_coverage_check = false` must also be set. In the setup used in [this](https://github.com/idaholab/moose/discussions/26941) example, also `material_coverage_check = false` must be set.

I assume that the usual requirements for the inactive domain are as follows (open for discussion):

- Nothing should happen to the elements in the inactive domains (no numerical instabilities, etc.)
- Elements in the inactive domain should not add as less weight as possible to the memory and/or CPU consuption.
- Elements from an inactive subdomain can be reactivated at the user's request.

Now my questions start, all dealing on what are the best practice to define these inactive domains:

1.  Is it really neccessary to define kernels in these inactive domains? I assume these kernels add some computational weight?
2. If kernels are to be added, what is the "best" kernel to add? Diffusion? MatDiffusion? From the description, [NullKernel](https://mooseframework.inl.gov/source/kernels/NullKernel.html) seems to be aimed at tests and debugging and not for "production".
3. Is there a way to avoid setting `kernel_coverage_check = false` and `material_coverage_check = false` while keeping the inactive subdomains lightweight? I'd love Moose to perform these checks in my "active" subdomains,
4. If the answer to point 3 above would be "no", what would be an acceptable solution as PR? A Parameter `skip_kernel_coverage_check_on = 'list of subdomains'` in [Problem] (same for `skip_material_coverage_check_on`)?

Kind regards,
Jörg

**Comment by GiudGiud** [comment, marked as answer]
> Nothing should happen to the elements in the inactive domains (no numerical instabilities, etc.)

There should be either no equations there because the variables are block-restricted away from them
OR
Only trivial equations (dc/dt = 0) if for some reason the variables need to still be there

> Elements in the inactive domain should not add as less weight as possible to the memory and/or CPU consuption.

Yes that seems fair. To do that, remove the variables and auxvariables from those subdomains

> Elements from an inactive subdomain can be reactivated at the user's request.

This seems fine if you move them back to the active subdomains

> Is it really neccessary to define kernels in these inactive domains? I assume these kernels add some computational weight?

No it's not. Just turn off the kernel_coverage_check

> If kernels are to be added, what is the "best" kernel to add? Diffusion? MatDiffusion? From the description, [NullKernel](https://mooseframework.inl.gov/source/kernels/NullKernel.html) seems to be aimed at tests and debugging and not for "production".

NullKernel is fine. A TimeDerivative is fine too

> Is there a way to avoid setting kernel_coverage_check = false and material_coverage_check = false while keeping the inactive subdomains lightweight? I'd love Moose to perform these checks in my "active" subdomains,

Not really. For now these checks are on/off. But you could change the checks to ignore the inactive subdomains. We would need new code to identify the inactive subdomains

> If the answer to point 3 above would be "no", what would be an acceptable solution as PR? A Parameter skip_kernel_coverage_check_on = 'list of subdomains' in [Problem] (same for skip_material_coverage_check_on)?

This seems fine. An alternative could be this:
```
skip_kernel_coverage_check = 'false / true / active'
```

**Reply by GiudGiud** [reply]
@hugary1995 or @dschwen  may want to pitch in

**Reply by hugary1995** [reply]
I just have a few things to add to what @GiudGiud suggested

First of all, there is an unfortunate terminology clash: I want to make it clear to whomever may read this thread in the future, the terms "active" and "inactive" used in element subdomain modifiers are NOT the same as those used in libmesh to denote h-refinement.

> Elements in the inactive domain should add as less weight as possible to the memory and/or CPU consumption.

If there is no equation system on inactive subdomains, the memory consumption is mostly just the mesh itself. However, things are complicated in parallel, as it depends on the mesh type, i.e., replicated vs distributed. See this discussion https://github.com/idaholab/moose/issues/27929 for more details.

Regarding kernel coverage check, there is no middle ground at the moment between "kernel coverage check everywhere" and "no kernel coverage check at all". As an alternative to @GiudGiud 's proposal, I'd recommend something like
```python
kernel_coverage_check = 'A B C'
```
where the user can specify a list of subdomains to perform the coverage check. The same goes to `material_coverage_check`. An empty list (default) would be equivalent to "all subdomains", in consistency with the `block` parameter in block-restrictable objects.

**Reply by jmeier** [reply]
Dear @hugary1995, 

Thank you for your answer and information provided.

On the parameter "kernel_coverage_check" as described by you: This would be opt-in. The opt-out method would be more fail-safe in the event that the user adds subdomains as the input file evolves over time. Therefore my suggestion was to use `skip_kernel_coverage_check_on = 'list of subdomains'` (same for `skip_material_coverage_check_on`)?

Jörg

**Reply by GiudGiud** [reply]
> An empty list (default) would be equivalent to "all subdomains", in consistency with the block parameter in block-restrictable objects.

Make sure 'false' is still an option to that parameter otherwise you'll have to patch 1000 inputs

> Therefore my suggestion was to use skip_kernel_coverage_check_on = 'list of subdomains' (same for skip_material_coverage_check_on)?

This is fine but ideally the default does the right thing. As in it skips empty / inactive subdomains. Which by specifying a list manually we wont have

**Reply by hugary1995** [reply]
I welcome any improvement. @Wendy-Ji since you have a lot of practical applications using this capability, do you have any comment to add here?

**Reply by jmeier** [reply]
> Make sure 'false' is still an option to that parameter otherwise you'll have to patch 1000 inputs

That's on me. I did not describe this hypothetical proposal in detail. I apologise for the confusion.

I'd like to keep `kernel_coverage_check = true/false` as is (while "true" is the default as it is today). The proposal is to just add the optional parameter `skip_kernel_coverage_check_on = 'list of subdomains'` while the default for this parameter is an empty list (no subdomains to be skipped). 

1. In case  `kernel_coverage_check = true` and `skip_kernel_coverage_check_on` is not set (or empty), the coverage check is to be done on all subdomains (as today). This is the *default*.
2. In case  `kernel_coverage_check = true` and `skip_kernel_coverage_check_on = 'list of some subdomains'`, the coverage check is to be done on all subdomains except the subdomains listed by the user.
3.  In case  `kernel_coverage_check = false` and `skip_kernel_coverage_check_on` is not set (or empty), the coverage check is to be done on no subdomain at all (as today).
4. In case  `kernel_coverage_check = false` and `skip_kernel_coverage_check_on = 'list of some subdomains'`, the coverage check is to be done on no subdomain at all (and skip_kernel_coverage_check_on is ignored).

The same with material coverage.

**Reply by jmeier** [reply]
Dear @hugary1995 

> First of all, there is an unfortunate terminology clash: I want to make it clear to whomever may read this thread in the future, the terms "active" and "inactive" used in element subdomain modifiers are NOT the same as those used in libmesh to denote h-refinement.

What would be the appropriate term in context of Moose?

**Reply by Wendy-Ji** [reply]
I agree with improving the `kernel_coverage_check` to having more options than all or nothing. While it would be more convenient to have an option to only check over active subdomains, it might be better for a general case to go with manually listing subdomains to include/exclude in the check. I'm not sure what other applications there are for skipping particular subdomains, but it's at least an option. 

I've also found I run into `boundary_restricted` checks because of boundaries defined over the inactive subdomains, but since that's not checking subdomains and is instead checking variables over the boundaries, that's probably a different problem.

For the proposed parameter, @grunerjmeier , I don't think it's necessary to have two parameters controlling the `kernel_coverage_check`. The 4 cases listed aren't really separate. Cases 1 and 2 only use `skip_kernel_coverage_check_on` to see what subdomains to skip, while Cases 3 and 4 only use `kernel_coverage_check=false` to skip all the subdomains. It doesn't seem like there's 4 different scenarios that would need two parameters. 

Either having just a `kernel_coverage_check` with `true|false|list of subdomains to check`, or `skip_kernel_coverage_check` with `false|true|list of subdomains to skip`, would cover the scenarios of: checking all the subdomains, checking none of the subdomains, and checking particular subdomains, respectively. 

Also, since adding subdomains over time could introduce either more active or inactive subdomains, I'm not sure I see the advantage of `skip_kernel_coverage_check` over `kernel_coverage_check`. The only argument I would make is that listing subdomains to check would let you keep the current `kernel_coverage_check` parameter mostly as is, and then the third option can be added in.

**Reply by hugary1995** [reply]
One issue with that is, given `kernel_coverage_check = false`, how to distinguish between the boolean `false` and a subdomain named "false".

**Reply by GiudGiud** [reply]
we can always error if the Mesh contains a subdomain called false and that parameter is set

**Reply by jmeier** [reply]
**On `true|false|list of subdomains to check`**
Let's look at this `kernel_coverage_check = true|false|list of subdomains to check` from another perspective: What are the naming conventions for subdomains? Which characters are permitted in the names of subdomains and materials? Which are not? Can we use one of the non-permitted characters as an escape character? e.g. `**all**` or `**none**`. 
However, this approach does not feel like the usual Moose design pattern. And this would inavlidate existing input files with `kernel_coverage_check = true` or `kernel_coverage_check = false`.

**Better term "Inactive Subdomains"**
As noted by @hugary1995 [here](https://github.com/idaholab/moose/discussions/28092#discussioncomment-10012536), the term "inactive" gives an unfortunate terminology clash. What would be the appropriate term in context of Moose?

**Opt-In or Opt-Out**
From the synatax `kernel_coverage_check = 'list of subdomains to check'` I'd expect that the subdomains given in this list are to be checked for kernel coverage (opt-in). As also mentioned by others, there might be a *lot* of subdomains in a model. This would burden the user with a lot of typing if he only wanted to exclude just one subdomain from checking.
Furthermore, I suspect that the number of "inactive" subdomains in most models is smaller than the number of "active" subdomains. If this is the case, enumerating the subdomains not to be checked would be less time-consuming than enumerating the subdomains to be checked.

**Reply by GiudGiud** [reply]
Hello

> Which characters are permitted in the names of subdomains and materials

they are both std::string underneath and I suspect " " and also "=" are the major disallowed ones.
You can use regex in moose with *, . etc but you'll need to parse the parameter in the code, it's not something SubdomainName (or vector of) parameters do for you.

> And this would inavlidate existing input files with kernel_coverage_check = true or kernel_coverage_check = false.

why? If true/false are still an option we're good

> Better term "Inactive Subdomains"

Inactive subdomains is not bad. The active/inactive technology for h-refinement is for elements not for subdomains. 
Alternatives could be idle/dormant/inert but all of these are not amazing.
Besides active/inactive is also already used to declare objects in the input file to be active or inactive

> As also mentioned by others, there might be a lot of subdomains in a model. This would burden the user with a lot of typing if he only wanted to exclude just one subdomain from checking.

I agree with that. 
I think having two parameters:
```
kernel_coverage_check = true  #(default, could be omitted)
skip_kernel_coverage = 'my-inactive-subdomain'
```
is a very good solution

**Reply by Wendy-Ji** [reply]
Sorry, I didn't realise that different variable types were hard to put together in one parameter. In that case, a second parameter is better. I'd say for all the applications I've been using so far at least, that the number of inactive subdomains is significantly higher than the active subdomains, but it's also not that much effort to set up a list of all the subdomains to skip, so `skip_kernel_coverage` is fine.

**Reply by jmeier** [reply]
> I agree with that. I think having two parameters:
> ```
> kernel_coverage_check = true  #(default, could be omitted)
> skip_kernel_coverage = 'my-inactive-subdomain'
> ```

If we having two parameters, shall we use an enum for `kernel_coverage_check `?
```
kernel_coverage_check = true|false|skip|exclusive     #(default is "true")
kernel_coverage_blocks = 'list of subdomains' 
```

The behaviour could be:

- For `kernel_coverage_check=true` the parameter `kernel_coverage_blocks` must not be set and all subdomains are to be checked. This is the deault.
- For `kernel_coverage_check=false` the parameter `kernel_coverage_blocks` must not be set and no subdomains are to be checked.
- For `kernel_coverage_check=skip` subdomains given in the parameter `kernel_coverage_blocks` are not checked. All other subdomains will be checked.
- For `kernel_coverage_check=exclusive` only subdomains given in the parameter `kernel_coverage_blocks` are checked.

Would this be an better option? Do you see better terms than "skip" and "exclusive"?

With this solution, all existing input files would be still valid and work as before. In addition, the case mentioned by @Wendy-Ji is also covered if there are many more inactive than active subdomains.

**Reply by GiudGiud** [reply]
skip_list and only_list ?

**Reply by jmeier** [reply]
Shall we rename `kernel_coverage_blocks` to `kernel_coverage_block_list` in this case?
```
kernel_coverage_check = true|false|skip_list|only_list     #(default is "true")
kernel_coverage_block_list = 'list of subdomains' 
```

**Reply by jmeier** [reply]
On the things discussed on the kernel and material coverage checks:
I created issue #28117 and PR #28118 proposing an implementation of the ideas here.
Thanks for your support!

**Reply by GiudGiud** [reply]
@jmeier would you happen to have a good figure to show the use of these subdomain modifiers?
I am presenting on new systems in MOOSE (here mesh modifiers) at ANS Winter 2024 and could use it. I would credit you

**Reply by jmeier** [reply]
@GiudGiud: let me check...

**Reply by jmeier** [reply]
Dear @GiudGiud, please check your inbox. Please do not hesitate to contact me if you need something changed.

**Reply by GiudGiud** [reply]
This is really cool work, thank you!
If you get the chance this would be worth presenting here
https://inl.gov/mooseworkshop2025/
